src/core/macro.scala

```diff
@@ -259,12 +281,223 @@ object Macro:
+            field.name -> field.annotations.filter(filterAnnotation)
+        }
+
+      def groupTermsByName(tss: List[(String, List[Term])]): List[(String, List[Term])] = 
+        tss
+          .groupBy { case (name, _) => name }
+          .toList
+          .map { case (name, l) => name -> l.flatMap(_._2) }
+
+      def selectTermsByName(tss: List[(String, List[Term])], name: String): List[Term] = 
+        tss
+          .toMap
+          .getOrElse(name, Nil)
+      
+      def filterAnnotation(annTerm: Term): Boolean =
+        annTerm.tpe.typeSymbol.maybeOwner.isNoSymbol ||
+          annTerm.tpe.typeSymbol.owner.fullName != "scala.annotation.internal"
+
+      def getAnnsDict(typeRepr: TypeRepr): List[(String, List[Term])] = 
+        val annTerms: List[(String, List[Term])] = 
+          (fromConstructor(typeRepr.typeSymbol) ++ fromDeclarations(typeRepr.typeSymbol))
+            .filter {case (_, ts) => ts.nonEmpty}
+        groupTermsByName(annTerms)
+
+      def getInheritedAnnsDict(typeRepr: TypeRepr): List[(String, List[Term])] = 
+        val annTerms: List[(String, List[Term])] = 
+          typeRepr.baseClasses
+          .filterNot(isObjectOrScala)
+          .collect {
+            case s if s != typeRepr.typeSymbol =>
+              (fromConstructor(s) ++ fromDeclarations(s)).filter {
+                case (_, anns) => anns.nonEmpty
+              }
+            }
+          .flatten
+        groupTermsByName(annTerms)
+
+      def accumulateTypeAnnotations(typeRepr: TypeRepr): List[Term] = typeRepr match
+        case AnnotatedType(inner, ann) => ann :: accumulateTypeAnnotations(inner)
+        case _                         => Nil
+
+      def getTypeAnnsDict(typeRepr: TypeRepr): List[(String, List[Term])] = 
+        typeRepr
+          .typeSymbol
+          .caseFields
+          .map { field => 
+            val tpeRepr = field.tree match
+              case v: ValDef => v.tpt.tpe
+              case d: DefDef => d.returnTpt.tpe
+            
+            field.name -> accumulateTypeAnnotations(tpeRepr).filter { a =>
+              a.tpe.typeSymbol.maybeOwner.isNoSymbol ||
+              a.tpe.typeSymbol.owner.fullName != "scala.annotation.internal"
+            }
+
+          }
+          .filter(_._2.nonEmpty)
+
+      def getDefaultValue[P: Type](dict: Map[String, Expr[Any]], name: String): Expr[CallByNeed[Option[Any]]] = 
+        dict.get(name) match 
+          case Some(e) => '{ new CallByNeed(() => Some( ($e).asInstanceOf[P] )) }
+          case None => '{ new CallByNeed(() => None) }
+
+      extension (term: Term)
+        def asCallByNeed(tpe: TypeRepr): Term = 
+          Select(
+            '{CallByNeed}.asTerm, 
+            TypeRepr.of[CallByNeed.type].termSymbol.declaredMethod("apply").head
+            )
+            .appliedToType(tpe)
+            .appliedTo(term)
+
+      extension (tpe: TypeRepr)
+        def defaultValues: Map[String, Expr[Any]] = 
```

Comment:
maybe we can extract some common code with `defaultValue[T]` above?

Reply:
Refactored.